src/main/BlockPool.cpp

```diff
@@ -0,0 +1,863 @@
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include "rho/BlockPool.hpp"
+
+#include <assert.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+typedef std::uint64_t u64;
+typedef std::uint32_t u32;
+using std::function;
+
+#ifdef ALLOCATION_CHECK
+typedef std::map<void*, void*> allocation_map;
+static allocation_map allocations;
+
+static bool iterating = false;
+
+static void add_to_allocation_map(void* allocation, size_t size);
+static void remove_from_allocation_map(void* allocation);
+static void* lookup_in_allocation_map(void* tentative_pointer);
+#endif
+
+static void allocerr(const char* message) {
+  fprintf(stderr, "ERROR: %s\n", message);
+  abort();
+}
+
+using std::size_t;
+
+// Tracking heap bounds for fast rejection in pointer lookup.
+void* heap_start = reinterpret_cast<void*>(UINTPTR_MAX);
+void* heap_end = reinterpret_cast<void*>(0);
+
+// Hash bucket for the sparse block table and free lists.
+// The two lowest bits of the data pointer are flags indicating
+// if the hash bucket contains a superblock (2) and whether this
+// hash bucket is the first representing that allocation (1).
+struct HashBucket {
+  uintptr_t data;
+  unsigned size;
+};
+
+struct Superblock;
+
+// Freelists are used to store superblock nodes and large allocations.
+// For large allocations the block and superblock members are unused.
+struct FreeListNode {
+  FreeListNode* next;
+  u32 block;  // Used only for superblock free nodes.
+  Superblock* superblock;  // Used only for superblock free nodes.
+};
+
+// Superblocks are used to allocate small and medium sized objects.
+// The bitset tracks which blocks are currently allocated.
+// For small objects block_size is the actual object size.
+// For medium objects block_size is the 2-log of the object size.
+struct Superblock {
+  u32 block_size;
+  u32 next_untouched;
+  FreeListNode* free_list;
+  u64 free[];  // Free bitset.
+
+  Superblock(u32 block_size, unsigned bitset_entries):
+      block_size(block_size),
+      next_untouched(0),
+      free_list(nullptr) {
+    // Here we mark all bitset entries as free, later we don't have to do
+    // precise range checking while iterating allocated blocks when
+    // the number of blocks is not evenly divisible by 64.
+    for (int i = 0; i < bitset_entries; ++i) {
+      free[i] = ~0ull;
+    }
+  }
+};
+
+unsigned sparse_bits = 16;
+unsigned num_sparse_buckets = 1 << sparse_bits;
+unsigned hash_mask = num_sparse_buckets - 1;
+
+// The low pointer bits are masked out in the hash function.
+#define LOW_BITS (19)
+#define MAX_COLLISIONS (6)
+#define MAX_REBALANCE_COLLISIONS (3)
+
+// Computes the hash key for a pointer.
+static unsigned hash_ptr(uintptr_t ptr, unsigned hash_mask) {
+  unsigned low = (ptr << 4) & 0xFFFFFFFF;
```

Comment:
Yes, the argument has already been right-shifted. This left-shift is used to spread out the hash keys, so that adjacent objects are not sequential in the hash table. This leaves some room for collision buckets between the hash keys.